### PR TITLE
Add pulse gate pass

### PR DESCRIPTION
**Pull request by nkanazawa1989**
### Summary
This PR improves handling of pulse gates.

Current work flow

```python

# write custom gate
with pulse.build(name="my_gate") as x_gate:
    ...

circ1 = QuantmCircuit(1)
circ1.x(0)
circ1.delay(10, 0)
circ1.measure_all()

circ2 = QuantmCircuit(1)
circ2.x(0)
circ2.delay(15, 0)
circ2.measure_all()

# we need to repeat this line for all circs
circ1.add_calibration(gate="x", qubits=(0,), schedule=x_gate)
circ2.add_calibration(gate="x", qubits=(0,), schedule=x_gate)

circs = transpile([circ1 + circ2], backend)
backend.run(circs)
```

With new pass

```python
with pulse.build(name="my_gate") as x_gate:
    ...

backend.defaults().instruction_schedule_map.add(instruction="x", qubits=(0,), schedule=x_gate)

circ1 = QuantmCircuit(1)
circ1.x(0)
circ1.delay(10, 0)
circ1.measure_all()

circ2 = QuantmCircuit(1)
circ2.x(0)
circ2.delay(15, 0)
circ2.measure_all()

# instmap is automatically extracted from backend and x gate schedule is overridden
circs = transpile([circ1 + circ2], backend)
backend.run(circs)
```

This may benefit the calibration framework in Qiskit Experiment.

### Details and comments

The simple trick enabling this workflow is the schedule metadata "publisher" (name TBD, welcome suggestions). Note that schedules in the instmap that are provided by the backend are always registered by `PulseDefaults` constructor. This PR adds the publisher metadata of backend to the schedule registered by this. 

https://github.com/Qiskit/qiskit-terra/blob/9b48476cbaf5c4742f67e1ff6ce2546ba5fad01f/qiskit/providers/models/pulsedefaults.py#L204-L208

If `InstructionScheduleMap.add` is called by users, this will implicitly adds metadata of client (since we assume such schedule doesn't have metadata). This enables the pass to distinguish user-defined calibration from the backend provided calibrations. Thanks to this logic, we can selectively add user defined calibrations to the circuit.

TODO
- [x] Finalize the mechanism to distinguish backend/user-defined cals
- [x] Write unittest
- [x] Write reno

### CODE REVIEW

File: qiskit/transpiler/passes/calibration/pulse_gates.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+# that they have been altered from the originals.
+
+"""Add user provided calibrations."""
+
+from qiskit.dagcircuit import DAGCircuit
+from qiskit.pulse.instruction_schedule_map import InstructionScheduleMap, CalibrationPublisher
+from qiskit.transpiler.basepasses import TransformationPass
+
+
+class PulseGates(TransformationPass):
```
Comment (eggerdj) [inline qiskit/transpiler/passes/calibration/pulse_gates.py:20]: This may have some overlap with the following class: https://github.com/Qiskit/qiskit-terra/blob/9b48476cbaf5c4742f67e1ff6ce2546ba5fad01f/qiskit/transpiler/passes/scheduling/calibration_creators.py#L37-L69
is it possible to reduce this overlap by having a common base class?

File: qiskit/transpiler/preset_passmanagers/level0.py [hunk 2]
```diff
@@ -47,6 +47,7 @@
 from qiskit.transpiler.passes import ASAPSchedule
 from qiskit.transpiler.passes import AlignMeasures
 from qiskit.transpiler.passes import ValidatePulseGates
+from qiskit.transpiler.passes import PulseGates
```
Comment (eggerdj) [inline qiskit/transpiler/preset_passmanagers/level0.py:50]: Is this something that we want to have in the optimization levels? Perhaps #6403 might also be relevant here.

File: qiskit/transpiler/passes/calibration/pulse_gates.py (hunk 1 above)
Comment (nkanazawa1989) [reply qiskit/transpiler/passes/calibration/pulse_gates.py]: Good suggestion. see https://github.com/Qiskit/qiskit-terra/pull/6759/commits/94210f79ea4622f8f6e168064fca5097e3504415

File: qiskit/transpiler/preset_passmanagers/level0.py (hunk 2 above)
Comment (nkanazawa1989) [reply qiskit/transpiler/preset_passmanagers/level0.py:50]: `FullPassManager` would be great to manage this new pass. I think we can later switch to this framework.

**Comment by nkanazawa1989** [comment]
Thanks @eggerdj for the reviewing. I think this pass is necessary for all optimization levels, because calibration update and circuit optimization are separate context. For example, user should be able to create custom x gate, and use that gate with every optimization levels. Basically this is the same as writing :
```python
non_opt_circ = QuantumCircuit(nq)
...
non_opt_circ.add_calibration("x", (0,), schedule=x_gate)

opt0 = transpile(non_opt_circ, backend, optimization_level=0)
opt1 = transpile(non_opt_circ, backend, optimization_level=1)
opt2 = transpile(non_opt_circ, backend, optimization_level=2)
opt3 = transpile(non_opt_circ, backend, optimization_level=3)
```
however here we want to avoid explicitly writing `add_calibration` for each circuit. If we calibrate all basis gates, it would be very annoying to manually overriding all of quantum gates in the circuit.

**Comment by ajavadia** [comment]
There are two things that make me feel a bit uneasy about this approach:

1- The idea of changing `backend.defaults()`, only to be able to leverage `transpile(circ, backend)` later, is inconsistent with the rest Qiskit (I'm not aware of backend being changed elsewhere). 
In `transpile()`, the idea is that all info will be pulled from `backend`, but they can be over-ridden by other arguments (e.g. basis_gates, coupling_map, instruction_durations). So if I want to change the durations I don't change those of the backend, I supply my own `instruction_durations` like this:

```py
from qiskit.transpiler import InstructionDurations
durations = InstructionDurations.from_backend(backend)
durations.update('cx', (0, 1), 200, 'ns')
transpile(circ, backend, durations)  # this will over-ride the backend's durations
```

2- Is it really a big deal to write `add_calibration` for the circuits? If there are multiple calibrations you can put the in a `add_calibrations()` function and just write one line per circuit. This way at least it's explicit what's happening.

**Comment by nkanazawa1989** [comment]
Thanks Ali for the comments :)

> 1- The idea of changing backend.defaults(), only to be able to leverage transpile(circ, backend) later, is inconsistent with the rest Qiskit (I'm not aware of backend being changed elsewhere).

I don't remember where I saw this, but if I recall correctly 
- `backend.configurations`: immutable property.
- `backend.properties`: property that can be updated with calibration.
- `backend.defaults`: ptoperty that can be updated with calibration, and user _can_ override.

According to this [API docs](https://qiskit.org/documentation/stubs/qiskit.providers.models.PulseDefaults.html#qiskit-providers-models-pulsedefaults), `defaults` is

> Description of default settings for Pulse systems. These are instructions or settings that may be good starting points for the Pulse user. The user may modify these defaults for custom scheduling.

so overriding backend defaults is not quite unexpected usage. Note that I added `inst_map` argument to transpiler, so you can still override this with user provided argument.

> 2- Is it really a big deal to write add_calibration for the circuits?

Yes. With above code example it's not quite easy to understand the benefit of this implementation (you can still write `add_calibration` manually). However, this change will drastically improves user experience in qiskit-experiment and other qiskit modules that may use custom gate.

Let's assume following workflow. Here we want to calibrate SX gate and evaluate this with randomized benchmarking. Please compare the pseudo codes below.

in current workflow
```
define calibration

# run set of calibrations to define basis_gates
for cal_exp in calibration_experiments:
   result = cal_exp.run(backend)
   calibrations.update(result)

import rb
circuits = transpile(rb.circuits(backend), backend)  # we need to insert hook to add cals
for circ in circuits: 
   for q in circ.qubits:
     circ.add_calibration("sx", (q,), calibrations.get_schedule("sx", (q, )))

backend.run(circs)
```

what this PR proposes
```
define calibration

# run set of calibrations to define basis_gates
for cal_exp in calibration_experiments:
   result = cal_exp.run(backend)
   calibrations.update(result)

backend = calibrations.export()  # instmap is overridden with schedule with metadata

import rb
rb.run(backend)
```

The RB is a part of experiment so we can implement some calibration attachment logic in experiment. However, once we try to use calibrated gate for other expeirment, i.e. we may want to calibrate custom basis gates, such as [CCR gate](https://arxiv.org/abs/2103.00024) that implements iSWAP, and we may want use this for VQE. However, in this case, we need to write the hook shown in above example. This is really hurts user experience.

**Comment by ajavadia** [comment]
As commented in #6825 I think this approach is fine. 

This is a bit of me thinking out loud: I don't quite like this process of discovering which cals are supplied by the backend and which cals are supplied by the user seems a bit hacky (relying on 'metadata'). 

If I understand correctly you are doing this to attach user cals to the payload (`circ.add_calibration`). What if we change the payload model a bit and make *all* cals attached to the circuit? i.e. the payload contains instructions and all cals for those instructions. This makes the payload a bit larger, but I don't think by much if we're using parametric pulses. The advantage is that the payload is self-contained. Right now we have a frontend compiler and a backend compiler which means we can send the following 3 types of circuits:
- circuit with physical gates/qubits
- circuit with timing
- circuit with pulses

This is a bit arbitrary IMO. We could even go one level higher and ask backend compiler do more compilation --- so we can even send a circuit with toffolis as payload (e.g. ionq does this). I think a better philosophy is to make the frontend compiler create the actual physical circuit, which inevitably means resolving timing and also attaching pulse definitions for all the gates. This type of payload is then consistent with qasm3 as well. 

So even though the backend advertises its gate cals, we still include the subset of cals relevant to our circuit in the payload. It is then very explicit what the payload is doing (even if someone inspects it outside of the context of the backend).

**Comment by nkanazawa1989** [comment]
> The advantage is that the payload is self-contained

I like the idea. Since usually backends don't define many kind of basis gates, the overhead of attaching all gate defs is not quite large. However this sounds like we need matrix-like configuration rather than current `optimization_level`, namely

```
| opt level   | 0 | 1 | 2 |
| target code |   |   |   |
|-------------+---+---+---| 
| qasm sim    | A | B | C |   circuit with physical gates/qubits (1)
| backend     | D | E | F |   circuit with timing + circuit with pulses (2)
```

because generation of lower layer target code (timing/gate-attached) may hurt simulator performance. This is rough idea, but we may still need to separate `circuit with timing` and `circuit with pulses` because some circuit simulation with special noise models (e.g. T1/T2/static-ZZ simulation) require scheduled circuit. I'm not sure how such simulations are sensitive to performance issue, but I know (1) is very sensitive to this problem since there are many competitors.

Anyhow unifying real backend execution pass to (2) seems to make the chain cleaner.

> but I don't think by much if we're using parametric pulses

This is not true. In the instruction schedule map, pulse instructions are stored with assigned parameter because `InstructionScheduleMap` doesn't have a separate parameter table. Maybe this is worth considering since current implementation is designed before we have parametric pulses.

Another thing I don't quite understand is the behavior of backend compiler when calibration is attached to the circuit. If the compiler always recognizes the gate def entry as a custom pulse, it may duplicate the waveform data in the waveform memory, and eventually we could increase the risk of getting waveform memory overflow error.

**Comment by ajavadia** [comment]
> generation of lower layer target code (timing/gate-attached) may hurt simulator performance

Simulator backends shouldn't have cals in them, so I don't think any cals get attached anyway.
They also don't report timing for gates, so again scheduling wouldn't be done.
But if it's a simulator that mimicks a device, then I think it's important to include e.g. delays in scheduling as that affects the noise model.

> pulse instructions are stored with assigned parameter

Yeah definitely we should make those cals leverage parameterized pulses

**Comment by taalexander** [comment]
> What if we change the payload model a bit and make all cals attached to the circuit

I would be ok with this for a "pulse-level" payload.

We should not do this for all jobs (not sure if was being suggested) as:
- This has a lot of duplication and will increase the size of the payload (most calibrations are shared across a job)
- The pulse gate path is currently slower than the circuit only path and we will hurt performance on the backend for most users. While this will eventually be resolved, it will not be until the full release of QASM3 support.
- Not all users/backends support pulse gates and so this information is redundant/not available.

Another option would be to supply the override calibrations in a separate location and pull the override calibrations from a separate location. This would enable us to treat everything in that location as something that should be attached to the payload.

Personally, I have no problem with Backends being treated as mutable compilation targets. In the long run, I think they should even store/configure the passes that they will apply during thranspilation.

How do you determine whether to insert calibrations or not in transpile? Is there some option?

**Comment by nkanazawa1989** [comment]
Thanks Ali, Thomas,

Sounds like these comments are somewhat conflicting:

> What if we change the payload model a bit and make all cals attached to the circuit

> They also don't report timing for gates, so again scheduling wouldn't be done.

Perhaps we need tell the transpiler the _level_ of compilation? i.e. `simulation` / `hardware`. this makes me think we need different quantum circuit classes depending on the stage of compilation.

> supply the override calibrations in a separate location

If QASM3 code is generated per job we could serialize the instruction schedule map itself without attaching instructions to each circuit. In this case, how we can distinguish global `defcal` from one local to a circuit?

>How do you determine whether to insert calibrations or not in transpile? Is there some option?

This is implicitly done by leveraging the metadata in current design. Because the inst map instance (attached to backend object) is populated with `backend-calibrated` schedules, it doesn't contain any user provided gate unless we explicitly `instmap.add` schedule. So the pulse gate pass checks the metadata of all gate schedules appearing in the circuit. By default, there is no `user-provided` schedule in the map and no calibration is attached. This could be a compute-heavy pass and maybe we should be able to enable/disable. On the other hand, this mean we always need to manage that flag and pass it down to the compilation chain -- this is actually I wanted to avoid.

**Comment by taalexander** [comment]
> Perhaps we need tell the transpiler the level of compilation? i.e. simulation / hardware. this makes me think we need different quantum circuit classes depending on the stage of compilation.

> If QASM3 code is generated per job we could serialize the instruction schedule map itself without attaching instructions to each circuit. In this case, how we can distinguish global defcal from one local to a circuit?

This would require the ability to call defcals explicitly in the qasm3 and provide multiple calibration definitions for the same gate(which is planned). See https://github.com/Qiskit/openqasm/issues/244

**Comment by taalexander** [comment]
I am ok with the metadata approach. Regarding multiple circuit classes, I would rather prefer to unify the circuit/pulse within one object structure (IR). We could have different canonical phases/forms within this IR structure to denote the format at the current point in compilation (eg., mapped, scheduled, calibrated, etc.).

**Comment by nkanazawa1989** [comment]
Thank you Thomas. It sounds like we can finalize the approach to attach custom gates. I also added unittest. Now this PR is ready for the final review. Note that in the current implementation, the stage of compilation is not tracked. This means the pulse gate pass is triggered also for the circuit for, for example, QASM simulator.

To address some performance concerns, I added `has_custom_gate` method to the inst map to avoid running this heavy-compute pass when we don't have custom gates (https://github.com/Qiskit/qiskit-terra/pull/6759/commits/e6015f4328518c51e05d0231787c18e23b5e25ce). Because standard circuit simulation code doesn't override the backend inst map, the pulse gate pass is usually not triggered.

File: qiskit/transpiler/passes/calibration/creators.py [hunk 3]
```diff
@@ -0,0 +1,472 @@
...
+    ControlChannel,
+    DriveChannel,
+    GaussianSquare,
+)
+from qiskit.pulse.instruction_schedule_map import InstructionScheduleMap, CalibrationPublisher
+from qiskit.pulse.instructions.instruction import Instruction as PulseInst
+from qiskit.transpiler.basepasses import TransformationPass
+
+
+class CalibrationCreator(TransformationPass):
```
Comment (taalexander) [inline qiskit/transpiler/passes/calibration/creators.py:41]:
```suggestion
class CalibrationBuilder(TransformationPass):
```

This is more standard within compiler terminology.

File: qiskit/transpiler/passes/calibration/creators.py [hunk 4]
```diff
@@ -0,0 +1,472 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
```
Comment (taalexander) [inline qiskit/transpiler/passes/calibration/creators.py:5]: Rename module to `builders`

File: qiskit/transpiler/passes/calibration/creators.py [hunk 5]
```diff
@@ -0,0 +1,472 @@
...
+        rzx_theta = rzx_theta.insert(0, rot)
+        rzx_theta = rzx_theta.insert(0, Delay(cr.duration, DriveChannel(control)))
+
+        return rzx_theta
+
+
+class PulseGates(CalibrationCreator):
+    """Pulse gate adding pass.
+
+    This pass adds gate calibrations in InstructionScheduleMap to a quantum circuit.
```
Comment (taalexander) [inline qiskit/transpiler/passes/calibration/creators.py:416]:
```suggestion
    This pass adds gate calibrations from the supplied ``InstructionScheduleMap`` to a quantum circuit.
```

File: qiskit/transpiler/passes/calibration/creators.py [hunk 6]
```diff
@@ -0,0 +1,472 @@
...
+    In QASM3 [1], the gate calibration can be encapsulated as ``defcal`` entry that declares
+    the hardware implementation of ``gate`` in specified grammar, such as OpenPulse.
```
Comment (taalexander) [inline qiskit/transpiler/passes/calibration/creators.py:418]: Is this relevant to this pass? I think it may be removed.

File: qiskit/transpiler/passes/calibration/creators.py [hunk 7]
```diff
@@ -0,0 +1,472 @@
...
+    Because this map is a mutable object, end-user can provide configured backend to
+    execute the circuit with customized gate implementations.
+
+    This mapping object returns a schedule with "publisher" metadata which is an integer Enum
+    value representing who created the gate schedule.
+    If the gate schedule is provided by end-users, this pass attaches the schedule to
+    the DAG circuit as a ``defcal`` entry.
+
+    This pass allows users to easily override quantum circuit with custom gate definition
+    without directly dealing with those schedules.
```
Comment (taalexander) [inline qiskit/transpiler/passes/calibration/creators.py:431]:
```suggestion

    This pass checks each DAG circuit node and acquires a corresponding schedule from
    the instruction schedule map object that may be provided by the target backend.
    Because this map is a mutable object, the end-user can provide a configured backend to
    execute the circuit with customized gate implementations.

    This mapping object returns a schedule with "publisher" metadata which is an integer Enum
    value representing who created the gate schedule.
    If the gate schedule is provided by end-users, this pass attaches the schedule to
    the DAG circuit as a calibration.

    This pass allows users to easily override quantum circuit with custom gate definitions
    without directly dealing with those schedules.
```

File: qiskit/transpiler/passes/calibration/creators.py [hunk 8]
```diff
@@ -0,0 +1,472 @@
...
+                    channel=instruction.channel,
+                )
+        else:
+            raise QiskitError("RZXCalibrationBuilder only stretches/compresses GaussianSquare.")
+
+    def get_calibration(self, node_op: CircuitInst, qubits: List) -> Union[Schedule, ScheduleBlock]:
+        """Builds the calibration schedule for the RZXGate(theta) with echos.
+
+        Args:
+            node_op: Instruction of the RZXGate(theta). I.e. params[0] is theta.
```
Comment (taalexander) [inline qiskit/transpiler/passes/calibration/creators.py:189]: What happens if the input is not of type `RZXGate`?

File: qiskit/transpiler/preset_passmanagers/level0.py [hunk 9]
```diff
@@ -162,7 +164,10 @@ def _direction_condition(property_set):
 
     _direction = [GateDirection(coupling_map)]
 
-    # 7. Unify all durations (either SI, or convert to dt if known)
+    # 7. Add calibrations for pulse gates
+    _calibrations = PulseGates(inst_map=inst_map)
```
Comment (taalexander) [inline qiskit/transpiler/preset_passmanagers/level0.py:168]: Should this object only be created if required? Eg., it should be contained within the block on line 203? The same in the other pass managers.

File: releasenotes/notes/add-pulse-gate-pass-dc347177ed541bcc.yaml [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+deprecations:
+  - |
+    Transpiler pass reorganization regarding calibrations.
+    The import path for :class:`~qiskit.transpiler.passes.scheduling.calibration_creators.\
+    RZXCalibrationBuilder` and :class:`~qiskit.transpiler.passes.scheduling.\
+    calibration_creators.RZXCalibrationBuilderNoEcho` are deprecated.
+    The import path for :func:`qiskit.transpiler.passes.scheduling.rzx_templates.\
+    rzx_templates.rzx_templates` is also deprecated.
+
+    Use new import path under :mod:`~qiskit.transpiler.passes.calibration`.
```
Comment (taalexander) [inline releasenotes/notes/add-pulse-gate-pass-dc347177ed541bcc.yaml:94]:
```suggestion
---
features:
  - |
    A new transpiler pass group :mod:`~qiskit.transpiler.passes.calibration` has been added.
    This group consists of

    - :class:`~qiskit.transpiler.passes.calibration.creators.RZXCalibrationBuilder`
    - :class:`~qiskit.transpiler.passes.calibration.creators.RZXCalibrationBuilderNoEcho`
    - :func:`~qiskit.transpiler.passes.calibration.rzx_templates.rzx_templates`.
    - :class:`~qiskit.transpiler.passes.calibration.creators.PulseGates`

    ``RZXCalibrationBuilder``, ``RZXCalibrationBuilderNoEcho``, and ``rzx_templates`` are
    moved from :mod:`qiskit/transpiler/passes/scheduling`, whereas ``PulseGates``
    is the new pass that automatically extracts user-provided calibrations
    from the instruction schedule map and attaches the gate schedule to the
    given (transpiled) quantum circuit as a pulse gate.
    Any transpiler pass that interacts with the pulse gate feature will be grouped to
    :mod:`qiskit.transpiler.passes.calibration`.

    This pass is applied to all optimization levels from 0 to 3.
    No gate implementation is updated unless the end-user explicitly overrides
    the ``backend.defaults().instruction_schedule_map``.

    This pass saves users from individually calling :meth:`~qiskit.circuit.quantumcircuit.\
    QuantumCircuit#add_calibration` method for every circuit run on the hardware.

    Along with this change, a schedule was added to :class:`qiskit.pulse.instruction_schedule_map.\
    InstructionScheduleMap` and is implicitly updated with a metadata "publisher".
    Backend calibrated gate schedules have a special publisher kind to avoid overriding
    circuits with calibrations of already known schedules.
    Usually, end-users don't need to take care of this metadata as it is applied automatically.
    You can call
    :py:meth:`~qiskit.pulse.instruction_schedule_map.InstructionScheduleMap#has_custom_gate`
    to check if the map has custom gate calibration.

    See below code example to learn how to apply custom gate implementation
    for all circuits under execution.

    .. code-block:: python

      from qiskit.test.mock import FakeGuadalupe
      from qiskit import pulse, circuit, transpile

      backend = FakeGuadalupe()

      with pulse.build(backend, name="x") as x_q0:
          pulse.play(pulse.Constant(160, 0.1), pulse.drive_channel(0))

      backend.defaults().instruction_schedule_map.add("x", (0,), x_q0)

      circs = []
      for _ in range(100):
          circ = circuit.QuantumCircuit(1)
          circ.sx(0)
          circ.rz(1.57, 0)
          circ.x(0)
          circ.measure_active()
          circs.append(circ)

      circs = transpile(circs, backend)
      circs[0].calibrations  # This returns calibration only for x gate

    Note that the instruction schedule map is a mutable object.
    If you override one of the entries and use that backend for other experiments,
    you may accidentally update the gate definition.

    .. code-block:: python

      backend = FakeGuadalupe()

      instmap = backend.defaults().instruction_schedule_map
      instmap.add("x", (0, ), my_x_gate_schedule)

      qc = QuantumCircuit(1, 1)
      qc.x(0)
      qc.measure(0, 0)

      qc = transpile(qc, backend)  # This backend uses custom X gate

    If you want to update the gate definitions of a specific experiment,
    you need to first deepcopy the instruction schedule map
    and directly pass it to the transpiler.


deprecations:
  - |
    Transpiler pass reorganization regarding calibrations.
    The import path for :class:`~qiskit.transpiler.passes.scheduling.calibration_creators.\
    RZXCalibrationBuilder` and :class:`~qiskit.transpiler.passes.scheduling.\
    calibration_creators.RZXCalibrationBuilderNoEcho` are deprecated.
    The import path for :func:`qiskit.transpiler.passes.scheduling.rzx_templates.\
    rzx_templates.rzx_templates` is also deprecated.

    Use new import path under :mod:`~qiskit.transpiler.passes.calibration`.
```

**Review by taalexander — CHANGES_REQUESTED** [review]

File: qiskit/transpiler/passes/calibration/rzx_templates.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+
+def rzx_templates(template_list: List[str] = None):
+    """
+    Convenience function to get the cost_dict and
+    templates for template matching.
+    """
+
+    if template_list is None:
+        template_list = ["zz1", "zz2", "zz3", "yz", "xz", "cy"]
```
[the remark this answers is not in the document]
Comment (nkanazawa1989) [reply qiskit/transpiler/passes/calibration/rzx_templates.py:44]: Code is upgraded in https://github.com/Qiskit/qiskit-terra/pull/6759/commits/b4880bde8439c5b953a93cf7240567c86840d9f7

File: qiskit/transpiler/preset_passmanagers/level0.py (hunk 9 above)
Comment (nkanazawa1989) [reply qiskit/transpiler/preset_passmanagers/level0.py]: Code is upgraded in https://github.com/Qiskit/qiskit-terra/pull/6759/commits/4bfc87b41ae210aec03b79e4233816543a50f44e

File: qiskit/transpiler/passes/calibration/creators.py (hunk 8 above)
Comment (nkanazawa1989) [reply qiskit/transpiler/passes/calibration/creators.py]: This will not happen since the method is only called after `supported` method returns `True`. In the `RZX` pass family, the supported method is equivalent to `isinstance(node_op, RZXGate)`.

File: qiskit/transpiler/passes/calibration/creators.py (hunk 3 above)
Comment (nkanazawa1989) [reply qiskit/transpiler/passes/calibration/creators.py]: Code is upgraded in https://github.com/Qiskit/qiskit-terra/pull/6759/commits/e16f931a672ea55ef7804a76ace1f6362c18283d

**Review by taalexander — APPROVED** [review]
